tools/generator/program.go

```diff
@@ -26,399 +37,280 @@ import (
-	flag.BoolVar(&verbose, "verbose", false, "Print status messages as processing is done.")
-	flag.BoolVar(&wait, "wait", false, "Use this program to halt execution before the cleanup phase is entered.")
-	useDebug := flag.Bool("debug", false, "Include this flag to print debug messages as the program executes.")
-	rawVersion := flag.String("version", "0.0.0", "The version that should be stamped on the generated code. Common usage will be to report user agent.")
-
-	flag.Parse()
-
-	if help {
-		return
-	}
+// version should be set by the linker when compiling this program by providing the following arguments:
+// -X main.version=<version>
+//
+// If installing the generator in your machine, that means running the following the command:
+//   go install -ldflags "-X main.version=<version>"
+//
+// The reason this is not controlled in source, is to allow for the git commit SHA1 to be used as the
+// version of the string. To retrieve the currently checked-out git commit SHA1, you can use the command:
+//   git rev-parse HEAD
+var version string
 
-	version, err = semver.NewVersion(*rawVersion)
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "Could not read version \"%s\" because: %v", *rawVersion, err)
-		os.Exit(1)
-	}
+func main() {
+	start := time.Now()
 
-	debugWriter := ioutil.Discard
-	if *useDebug {
-		fmt.Fprintln(os.Stdout, "Turning on debug logger.")
-		debugWriter = os.Stderr
+	type generationTuple struct {
+		fileName     string
+		packageName  string
+		outputFolder string
 	}
-	debugLog = log.New(debugWriter, "[DEBUG]", log.Ltime)
 
-	optionalTools := []string{"gofmt", "golint"}
-	requiredTools := []string{"autorest", "git"}
+	literateFiles := collection.Where(targetFiles, func(subject interface{}) bool {
+		return strings.EqualFold(path.Base(subject.(string)), "README.md")
+	})
 
-	for _, tool := range optionalTools {
-		if _, err := exec.LookPath(tool); err != nil {
-			log.Printf("WARNING: Could not find \"%s\" usage of this tool will be skipped.", tool)
-		}
-	}
+	tuples := collection.SelectMany(literateFiles,
+		// The following function compiles the regexp which finds Go related package tags in a literate file, and creates a collection.Unfolder.
+		// This function has been declared this way so that the relatively expensive act of compiling a regexp is only done once.
+		func() collection.Unfolder {
+			const patternText = "```" + `\s+yaml\s+\$\(\s*tag\s*\)\s*==\s*'([\d\w\-\.]+)'\s*&&\s*\$\(go\)[\w\d\-\s:]*\s+output-folder:\s+\$\(go-sdk-folder\)([\w\d\-_\\/\.]+)\s+[\w\d\-\s:]*` + "```"
+
+			goConfigPattern := regexp.MustCompile(patternText)
+
+			// This function is a collection.Unfolder which takes a literate file as a path, and retrieves all configuration which applies to a package tag and Go.
+			return func(subject interface{}) collection.Enumerator {
+				results := make(chan interface{})
```

Comment:
can it be a `generationTuple` from the beginning?

Reply:
I'm afraid not :/

Reply:
That's not to say there isn't anything we could do to make it nicer! But the generationTuple type would need to be significantly modified.